### PR TITLE
Change `Call` format to 1.7 version

### LINKED ISSUE #125: [v2] Rename Call17 to Call and Call to Call16
`CALL_16` behaviour may be confusing, especially for new users. Say, one returns a scalar from a function, but receive an array in this place in the connector (not counting one extra array for multireturn support). We should be in-line with tarantool here and provide `CALL_17` as just `Call`.

This is a breaking change and I propose to do it after we'll resolve most of more annoying problems (such as support decimals, contexts, master discovery, SQL).

There is a more gentle way to do so: keep API that is offered by `import "github.com/tarantool/go-tarantool"` as is, implement the new one as `import "github.com/tarantool/go-tarantool/v2"` or so and replace the former with the latter in README and docs. Anyway, it requires a careful planning of v2, so I still propose to return here somewhere around end of Q1'2022.

Ideas? Suggestions?

-----

The general discussion about v2 features: #65.

### PR DESCRIPTION
**Pull request by AnaNek**
An incompatible change was introduced in
Tarantula 1.7 which was released in 2016.
This change is about a new binary protocol
command for CALL, which no more restricts
a function to returning an array of tuples
and allows returning an arbitrary MsgPack/JSON
result, including scalars, nil and void (nothing).
    
We should be in-line with tarantool here and
provide `Call17` as just `Call` and rename
current `Call` to `Call16`.
    
Closes #125

### CODE REVIEW

**Comment by DifferentialOrange** [comment]
> Tarantula 1.7

Quite an usual way to name our product

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Added
 
+- Rename `Call` to `Call16` and `Call17` to `Call` (#125)
```
Comment (DifferentialOrange) [inline CHANGELOG.md:13]: We definitely should add a big bold **Breaking:** here because it'll break all code with `Call`s: both for 1.6 and post-1.6 users.

File: const.go [hunk 2]
```diff
@@ -6,11 +6,11 @@ const (
 	ReplaceRequest   = 3
 	UpdateRequest    = 4
 	DeleteRequest    = 5
-	CallRequest      = 6 /* call in 1.6 format */
+	Call16Request    = 6 /* call in 1.6 format */
 	AuthRequest      = 7
 	EvalRequest      = 8
 	UpsertRequest    = 9
-	Call17Request    = 10
+	CallRequest      = 10 /* call in 1.7 format */
```
Comment (DifferentialOrange) [inline const.go:13]: I don't think there are many 1.7 users and it could be not obvious for someone that 1.8, 1.9, 1.10 and the whole 2.x family included here. It we want to write a comment, we should add something like "call in >1.7 format"

File: multi/multi.go [hunk 3]
```diff
@@ -321,18 +321,18 @@ func (connMulti *ConnectionMulti) Upsert(space interface{}, tuple, ops interface
...
 // arrays.
-func (connMulti *ConnectionMulti) Call(functionName string, args interface{}) (resp *tarantool.Response, err error) {
-	return connMulti.getCurrentConnection().Call(functionName, args)
+func (connMulti *ConnectionMulti) Call16(functionName string, args interface{}) (resp *tarantool.Response, err error) {
+	return connMulti.getCurrentConnection().Call16(functionName, args)
 }
 
-// Call17 calls registered Tarantool function.
+// Call calls registered Tarantool function.
 // It uses request code for Tarantool 1.7, so result is not converted
```
Comment (DifferentialOrange) [inline multi/multi.go:332]: Same here about specifying only 1.7 version in comment, We should re-check all the code about 1.7 references and rewrite them this way

**Review by DifferentialOrange — COMMENTED** [review]
Thank you for your patch! Everything seems okay except for some minor things I've pointed out in comments.

However, we definitely shouldn't merge it yet. This is a breaking change, not only by its concept, but also by its effect. There are a lot of Go-Tarantool applications that use Calls, including our PRs (for example, your #148). And each of them will be broken after merging this change.

So we should
* implement and collect all of our planned v2-changes (see https://github.com/tarantool/go-tarantool/issues/65);
* learn how to deal with packages major version tags.

The first task is related to our planning strategy, so I think there is no use to discuss it here. The second task requires some research, and for now I don't see an issue related to this specific task. Maybe we should solve it as a part of this PR or file a separate issue and assign someone.

File: CHANGELOG.md (hunk 1 above)
Comment (AnaNek) [reply CHANGELOG.md]: Done.

File: const.go (hunk 2 above)
Comment (AnaNek) [reply const.go]: Done.

File: multi/multi.go (hunk 3 above)
Comment (AnaNek) [reply multi/multi.go]: Done.

File: const.go [hunk 4]
```diff
@@ -6,11 +6,11 @@ const (
 	ReplaceRequest   = 3
 	UpdateRequest    = 4
 	DeleteRequest    = 5
-	CallRequest      = 6 /* call in 1.6 format */
+	Call16Request    = 6 /* call in 1.6 format */
 	AuthRequest      = 7
 	EvalRequest      = 8
 	UpsertRequest    = 9
-	Call17Request    = 10
+	CallRequest      = 10 /* call in >=1.7 format */
```
Comment (DifferentialOrange) [inline const.go:13]:
```suggestion
	CallRequest      = 10 /* call in >= 1.7 format */
```

File: multi/multi.go [hunk 5]
```diff
@@ -321,18 +321,18 @@ func (connMulti *ConnectionMulti) Upsert(space interface{}, tuple, ops interface
...
-func (connMulti *ConnectionMulti) Call(functionName string, args interface{}) (resp *tarantool.Response, err error) {
-	return connMulti.getCurrentConnection().Call(functionName, args)
+func (connMulti *ConnectionMulti) Call16(functionName string, args interface{}) (resp *tarantool.Response, err error) {
+	return connMulti.getCurrentConnection().Call16(functionName, args)
 }
 
-// Call17 calls registered Tarantool function.
-// It uses request code for Tarantool 1.7, so result is not converted
+// Call calls registered Tarantool function.
+// It uses request code for Tarantool >=1.7, so result is not converted
```
Comment (DifferentialOrange) [inline multi/multi.go:332]:
```suggestion
// It uses request code for Tarantool >= 1.7, so result is not converted
```

File: multi/multi.go [hunk 6]
```diff
@@ -375,18 +375,18 @@ func (connMulti *ConnectionMulti) UpdateTyped(space, index interface{}, key, ops
...
-func (connMulti *ConnectionMulti) CallTyped(functionName string, args interface{}, result interface{}) (err error) {
-	return connMulti.getCurrentConnection().CallTyped(functionName, args, result)
+func (connMulti *ConnectionMulti) Call16Typed(functionName string, args interface{}, result interface{}) (err error) {
+	return connMulti.getCurrentConnection().Call16Typed(functionName, args, result)
 }
 
-// Call17Typed calls registered function.
-// It uses request code for Tarantool 1.7, so result is not converted (though,
+// CallTyped calls registered function.
+// It uses request code for Tarantool >=1.7, so result is not converted (though,
```
Comment (DifferentialOrange) [inline multi/multi.go:386]:
```suggestion
// It uses request code for Tarantool >= 1.7, so result is not converted (though,
```

File: multi/multi.go [hunk 7]
```diff
@@ -429,18 +429,18 @@ func (connMulti *ConnectionMulti) UpsertAsync(space interface{}, tuple interface
...
-func (connMulti *ConnectionMulti) CallAsync(functionName string, args interface{}) *tarantool.Future {
-	return connMulti.getCurrentConnection().CallAsync(functionName, args)
+func (connMulti *ConnectionMulti) Call16Async(functionName string, args interface{}) *tarantool.Future {
+	return connMulti.getCurrentConnection().Call16Async(functionName, args)
 }
 
-// Call17Async sends a call to registered Tarantool function and returns Future.
-// It uses request code for Tarantool 1.7, so future's result will not be converted
+// CallAsync sends a call to registered Tarantool function and returns Future.
+// It uses request code for Tarantool >=1.7, so future's result will not be converted
```
Comment (DifferentialOrange) [inline multi/multi.go:440]:
```suggestion
// It uses request code for Tarantool >= 1.7, so future's result will not be converted
```

File: request.go [hunk 8]
```diff
@@ -96,21 +96,21 @@ func (conn *Connection) Upsert(space interface{}, tuple, ops interface{}) (resp
...
-	return conn.CallAsync(functionName, args).Get()
+// It is equal to conn.Call16Async(functionName, args).Get().
+func (conn *Connection) Call16(functionName string, args interface{}) (resp *Response, err error) {
+	return conn.Call16Async(functionName, args).Get()
 }
 
-// Call17 calls registered Tarantool function.
-// It uses request code for Tarantool 1.7, so result is not converted
+// Call calls registered Tarantool function.
+// It uses request code for Tarantool >=1.7, so result is not converted
```
Comment (DifferentialOrange) [inline request.go:108]:
```suggestion
// It uses request code for Tarantool >= 1.7, so result is not converted
```

File: request.go [hunk 9]
```diff
@@ -188,21 +188,21 @@ func (conn *Connection) UpdateTyped(space, index interface{}, key, ops interface
...
-	return conn.CallAsync(functionName, args).GetTyped(result)
+// It is equal to conn.Call16Async(functionName, args).GetTyped(&result).
+func (conn *Connection) Call16Typed(functionName string, args interface{}, result interface{}) (err error) {
+	return conn.Call16Async(functionName, args).GetTyped(result)
 }
 
-// Call17Typed calls registered function.
-// It uses request code for Tarantool 1.7, so result is not converted
+// CallTyped calls registered function.
+// It uses request code for Tarantool >=1.7, so result is not converted
```
Comment (DifferentialOrange) [inline request.go:200]:
```suggestion
// It uses request code for Tarantool >= 1.7, so result is not converted
```

File: request.go [hunk 10]
```diff
@@ -320,11 +320,11 @@ func (conn *Connection) CallAsync(functionName string, args interface{}) *Future
 	})
 }
 
-// Call17Async sends a call to registered Tarantool function and returns Future.
-// It uses request code for Tarantool 1.7, so future's result will not be converted
+// CallAsync sends a call to registered Tarantool function and returns Future.
+// It uses request code for Tarantool >=1.7, so future's result will not be converted
```
Comment (DifferentialOrange) [inline request.go:324]:
```suggestion
// It uses request code for Tarantool >= 1.7, so future's result will not be converted
```

**Review by DifferentialOrange — APPROVED** [review]
LGTM after minor fixes. Let's merge it when we will be ready to tag v2.

File: request.go [hunk 11]
```diff
@@ -307,10 +307,10 @@ func (conn *Connection) UpsertAsync(space interface{}, tuple interface{}, ops in
 	})
 }
 
-// CallAsync sends a call to registered Tarantool function and returns Future.
+// Call16Async sends a call to registered Tarantool function and returns Future.
 // It uses request code for Tarantool 1.6, so future's result is always array of arrays
```
Comment (oleg-jukovec) [inline request.go:311]: Maybe we should add the exact version of Tarantool, starting from which an user should use Call instead of Call16 and why?

https://github.com/tarantool/tarantool/blob/fea8c8e40611581d85ab5f96ea5a11e348c02247/src/box/lua/net_box.lua#L760-L765

It is not clear which method should be used in a new code. The user can choose Call16 because it supports more versions of Tarantool.

**Comment by DifferentialOrange** [comment]
Btw, we need to rebase this PR onto master since there are calls in connection_pool too.

File: const.go (hunk 4 above)
Comment (AnaNek) [reply const.go]: Done.

File: multi/multi.go (hunk 5 above)
Comment (AnaNek) [reply multi/multi.go]: Done.

File: multi/multi.go (hunk 6 above)
Comment (AnaNek) [reply multi/multi.go]: Done.

File: multi/multi.go (hunk 7 above)
Comment (AnaNek) [reply multi/multi.go]: Done.

File: request.go (hunk 8 above)
Comment (AnaNek) [reply request.go]: Done.

File: request.go (hunk 9 above)
Comment (AnaNek) [reply request.go]: Done.

File: request.go (hunk 10 above)
Comment (AnaNek) [reply request.go]: Done.

File: request.go (hunk 11 above)
Comment (AnaNek) [reply request.go:315]: Updated:

```golang
 // Deprecated since 1.7.4. This breaking change was needed to extend call() and eval()
 // API with per-requests options.
 func (conn *Connection) Call16Async(functionName string, args interface{}) *Future {
 	future := conn.newFuture(Call16Request)
```

Comment (oleg-jukovec) [reply request.go:315]: I think just "Deprecated since Tarantool 1.7.4" is enough here (you can add a link to the pull request in Tarantool, but this is not necessary).

The mention of Tarantool's eval() and call() and API can be confusing (an user can think that it's about go-tarantool API).

**Comment by Totktonada** [comment]
I have several comments and thoughts.

----

> An incompatible change was introduced in Tarantool 1.7 <...>

The protocol was extended in the backward-compatible way: `CALL_16` still works and have the same request type number as in tarantool 1.6. The new `CALL_17` request type was introduced with its own request type number.

The change in tarantool ([1.7.1-233-gd025d102f][1.7.1-233-gd025d102f]) is actually incompatible: net.box's :call() behaviour was changed. But the protocol was changed in the compatible way.

Since we speaking here in context of a connector and the binary protocol, I found the cited words confusing. I propose to reword it to something like '<...> the new `CALL_17` request was instroduced and the old `CALL_16` request was deprecated'.

----

> Tarantool 1.7 which was released in 2016

There were several 1.7 releases. 1.7.2 was released in 2016, 1.7.4 in 2017.

The change of the question lands into 1.7.2. I would say 1.7.2 everywhere instead of just 1.7, because 1.7.* releases are quite different.

BTW, the API documentation mentions 1.7.4. It looks like a mistake. The protocol change is 1.7.1-233-gd025d102f, it lands to 1.7.2.

----

> allows returning an arbitrary MsgPack/JSON result

JSON looks irrelevant here. I see, it is from tarantool 1.7.2 release notes, but I found it confusing rather than explanatory.

----

We recently discussed another incompatible change with @oleg-jukovec (msgpack v5 support). Oleg proposed to use build tags to choose between msgpack v2 and msgpack v5. Here we can do the same.

I imagine it as follows. Now we have `Call` (as `CALL_16`) and `Call_17`. We can do the following:

1. Add `Call_16`.
2. Add `const_call_16.go` and `const_call_17.go`, which will look like so:

   ```go
   //go:build !go_tarantool_call_17
   // +build !go_tarantool_call_17

   package tarantool

   const (
           CallRequest = 6 // CALL_16
   )
   ```

   and

   ```go
   //go:build go_tarantool_call_17
   // +build go_tarantool_call_17

   package tarantool

   const (
           CallRequest = 10 // CALL_17
   )
   ```
3. Drop the tag in the next major release and move the `CallRequest` constant
   back to `const.go` with the new value (10).

It is the raw idea and I would like to hear more opinions.

[1.7.1-233-gd025d102f]: https://github.com/tarantool/tarantool/commit/d025d102f47a5247b24027435c8ecf54c09c7201

File: request.go (hunk 11 above)
Comment (AnaNek) [reply request.go]: Done.

File: CHANGELOG.md [hunk 12]
```diff
@@ -15,6 +15,8 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Changed
 
+- Breaking change: rename `Call` to `Call16` and `Call17` to `Call` (#125)
```
Comment (Totktonada) [inline CHANGELOG.md:18]: *(I don't know, whether current patch version is ready for review: sorry if the comment is premature.)*

I would not call this change breaking: if a user does not set the build tag, the behaviour is exatly same as before. We just added `Call_16` method.

File: connection.go [hunk 13]
```diff
@@ -116,9 +116,9 @@ func (d defaultLogger) Report(event ConnLogKind, conn *Connection, v ...interfac
 //
 // ATTENTION: result argument for *Typed methods should deserialize from
 // msgpack array, cause Tarantool always returns result as an array.
-// For all space related methods and Call* (but not Call17*) methods Tarantool
+// For all space related methods and Call16* (but not Call*) methods Tarantool
```
Comment (Totktonada) [inline connection.go:119]: `Call` still works as `Call16` by default, so the replacement of `Call17` to `Call` looks premature. I would use explicit `Call16` and `Call17` in the documentation.

File: connection_pool/connection_pool.go [hunk 14]
```diff
@@ -266,8 +268,20 @@ func (connPool *ConnectionPool) Call(functionName string, args interface{}, user
 	return conn.Call(functionName, args)
 }
 
+// Call16 calls registered Tarantool function.
+// It uses request code for Tarantool 1.6, so result is converted to array of arrays.
+// Deprecated since 1.7.4.
```
Comment (Totktonada) [inline connection_pool/connection_pool.go:273]: I mentioned this above, but I'll remember just in case: the new IPROTO command (and the deprecation of the old one) was in 1.7.2, not 1.7.4. See https://github.com/tarantool/tarantool/releases/tag/1.7.2.

File: CHANGELOG.md (hunk 12 above)
Comment (AnaNek) [reply CHANGELOG.md]: Updated:

```diff
-- Breaking change: rename `Call` to `Call16` and `Call17` to `Call` (#125)
+- Add `Call16` method, support build tag `go_tarantool_call_17`
+  to choose behavior for `Call` method (#125)
```

File: connection_pool/connection_pool.go (hunk 14 above)
Comment (Totktonada) [reply connection_pool/connection_pool.go:273]: BTW, it also seems worthful to say 'Deprecated since tarantool 1.7.2', not just 1.7.2, because the latter could be misunderstood as the version of the connector.

**Review by Totktonada — APPROVED** [review]

File: call_16_test.go [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+	. "github.com/tarantool/go-tarantool"
+)
+
+func TestConnection_Call(t *testing.T) {
+	var resp *Response
+	var err error
+
+	conn := connect(t, server, opts)
+	defer conn.Close()
```
Comment (oleg-jukovec) [inline call_16_test.go:17]: I think we need such tests for ConnectionPool and ConnectionMulti. Looks like we won't test ConnectionPool.Call() and ConnectionMulti.Call().

File: CHANGELOG.md [hunk 16]
```diff
@@ -15,6 +15,9 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Changed
 
+- Add `Call16` method, support build tag `go_tarantool_call_17`
```
Comment (oleg-jukovec) [inline CHANGELOG.md:18]: Please add a description of the tag to [README.md#build-tags](README.md#build-tags).


You can copy-paste changes in [README.md#build-tags](https://github.com/tarantool/go-tarantool/pull/174/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) from #174 (but don't copy about go_tarantool_msgpack_v5) if you want. But I do not insist.

**Review by oleg-jukovec — APPROVED** [review]
Thank you for the pull request! ILGTM after resolving the notes.

File: const_call_16.go [hunk 17]
```diff
@@ -0,0 +1,8 @@
+//go:build !go_tarantool_call_17
+// +build !go_tarantool_call_17
+
+package tarantool
+
+const (
+	CallRequest = 6 // CALL_16
```
Comment (oleg-jukovec) [inline const_call_16.go:7]:
```suggestion
	CallRequest = Call16Request
```

File: const_call_17.go [hunk 18]
```diff
@@ -0,0 +1,8 @@
+//go:build go_tarantool_call_17
+// +build go_tarantool_call_17
+
+package tarantool
+
+const (
+	CallRequest = 10 // CALL_17
```
Comment (oleg-jukovec) [inline const_call_17.go:7]:
```suggestion
	CallRequest = Call17Request
```

File: call_16_test.go (hunk 15 above)
Comment (AnaNek) [reply call_16_test.go:17]: Done.

File: const_call_17.go (hunk 18 above)
Comment (AnaNek) [reply const_call_17.go]: Done.

File: connection_pool/connection_pool.go (hunk 14 above)
Comment (AnaNek) [reply connection_pool/connection_pool.go]: Updated.

File: connection.go (hunk 13 above)
Comment (AnaNek) [reply connection.go]: Fixed.

File: connection_pool/connection_pool.go [hunk 19]
```diff
@@ -266,8 +268,20 @@ func (connPool *ConnectionPool) Call(functionName string, args interface{}, user
 	return conn.Call(functionName, args)
 }
 
+// Call16 calls registered Tarantool function.
+// It uses request code for Tarantool 1.6, so result is converted to array of arrays.
+// Deprecated since tarantool 1.7.2.
```
Comment (oleg-jukovec) [inline connection_pool/connection_pool.go:273]: Let's change in all places:
```suggestion
// Deprecated since Tarantool 1.7.2.
```
becase we use `Tarantool` in the code:
```bash
$ grep "Tarantool 1" -r . | wc -l
48
```

**Comment by oleg-jukovec** [comment]
It would be better to make the commit message wider:
https://www.tarantool.io/en/doc/latest/dev_guide/developer_guidelines/#how-to-write-a-commit-message
`Wrap the body to 72 characters or so.`

Now it stands out a little among others commit messages.

File: connection_pool/connection_pool.go (hunk 19 above)
Comment (AnaNek) [reply connection_pool/connection_pool.go]: Updated.
